Fix for whitelist/blacklist checking for non-list iterables

terminalmage opened the pull request:
A recent PY3-compatibility fix was made to `check_whitelist_blacklist`.  The function was previously checking for the `__iter__` attribute to check if the blacklist and/or whitelist was a sequence of expressions or a single expression. However, the `str` type has the `__iter__` attribute in PY3 while it does not in PY2. To fix this, the check was changed to see if the blacklist and/or whitelist were `list` types. This however causes problems in `salt.daemons.masterapi` when we pass `set` types.

To fix this, we just check if the whitelist and/or blacklist are string types, rather than that they are _not_ some sort of sequence type.

terminalmage committed:
Fix for whitelist/blacklist checking for non-list iterables

A recent PY3-compatibility fix was made to `check_whitelist_blacklist`.
The function was previously checking for the `__iter__` attribute to
check if the blacklist and/or whitelist was a sequence of expressions
or a single expression. However, the `str` type has the `__iter__`
attribute in PY3 while it does not in PY2. To fix this, the check was
changed to see if the blacklist and/or whitelist were `list` types. This
however causes problems in `salt.daemons.masterapi` when we pass `set`
types.

To fix this, we just check if the whitelist and/or blacklist are string
types, rather than that they are _not_ some sort of sequence type.

terminalmage committed:
Fix comment in test

terminalmage committed:
Raise a TypeError when invalid input passed to check_whitelist_blacklist

rallytime commented:
@terminalmage There are some test failures here: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-cent7-py3/4479/

terminalmage commented:
re-run all

terminalmage committed:
Treat empty whitelist/blacklist as no whitelist/blacklist

terminalmage commented:
OK, test failures should be fixed now. The tests exposed a flaw in my fix, which I addressed and for which I added additional asserts in the new unit test added in this PR.
